src/patcher.c

```diff
@@ -0,0 +1,835 @@
+/*
+ * copy_ymm_handler_code
+ * This routine copies the code saving/restoring the YMM (256 bit wide )
+ * registers in an assembly wrapper template. Without this, the default
+ * code for saving/restoring the XMM (128 bit wide) registers stays
+ * in the generated code.
+ */
+static void
+copy_ymm_handler_code(unsigned char *asm_wrapper)
+{
+	memcpy(asm_wrapper + o_simd_save,
+	    &intercept_asm_wrapper_simd_save_YMM, simd_save_YMM_size);
+	memcpy(asm_wrapper + o_simd_restore,
+	    &intercept_asm_wrapper_simd_restore_YMM, simd_restore_YMM_size);
+}
+
+/*
+ * create_push_imm
+ * Generates a push instruction, that pushes a 32 bit constant to the stack.
+ */
+static void
+create_push_imm(unsigned char *push, uint32_t syscall_offset)
+{
+	push[0] = PUSH_IMM_OPCODE;
+	push[1] = (unsigned char)(syscall_offset % 0x100);
+	push[2] = (unsigned char)((syscall_offset / 0x100) % 0x100);
+	push[3] = (unsigned char)((syscall_offset / 0x10000) % 0x100);
+	push[4] = (unsigned char)((syscall_offset / 0x1000000) % 0x100);
+}
+
+/*
+ * create_movabs_r11
+ * Generates a movabs instruction, that assigns a 64 bit constant to
+ * the R11 register.
+ */
+static void
+create_movabs_r11(unsigned char *code, uint64_t value)
+{
+	unsigned char *bytes = (unsigned char *)&value;
+
+	code[0] = 0x49; /* movabs opcode */
+	code[1] = 0xbb; /* specifiy r11 as destination */
+	code[2] = bytes[0];
+	code[3] = bytes[1];
+	code[4] = bytes[2];
+	code[5] = bytes[3];
+	code[6] = bytes[4];
+	code[7] = bytes[5];
+	code[8] = bytes[6];
+	code[9] = bytes[7];
+}
+
+/*
+ * create_wrapper
+ * Generates an assembly wrapper. Copies the template written in
+ * intercept_template.s, and generates the instructions specific
+ * to a particular syscall into the new copy.
+ * After this wrapper is created, a syscall can be replaced with a
+ * jump to this wrapper, and wrapper is going to call dest_routine
+ * (actually only after a call to mprotect_asm_wrappers).
+ */
+static void
+create_wrapper(struct patch_desc *patch, void *dest_routine,
+			bool use_absolute_return,
+			const char *libpath)
+{
+	unsigned char *begin;
+
+	/* Create a new copy of the template */
+	patch->asm_wrapper = begin = next_asm_wrapper_space();
+	memcpy(begin, intercept_asm_wrapper_tmpl, tmpl_size);
+
+	/* Copy the prev/next instrucions, if they are copiable */
+	if (patch->uses_prev_ins) {
+		size_t length = patch->preceding_ins.length;
+		if (patch->uses_prev_ins_2)
+			length += patch->preceding_ins_2.length;
+
+		memcpy(begin + o_prefix, patch->syscall_addr - length, length);
+	}
+	if (patch->uses_next_ins) {
+		memcpy(begin + o_postfix,
+		    patch->syscall_addr + SYSCALL_INS_SIZE,
+		    patch->following_ins.length);
+	}
+
+	if (patch->syscall_offset > UINT32_MAX)
+		xabort(); /* libc larger than 2 gigabytes? wow */
+
+	/* the instruction pushing the syscall's address to the stack */
+	create_push_imm(begin + o_push_origin, (uint32_t)patch->syscall_offset);
+
+	create_movabs_r11(begin + o_mov_return_r11_no_syscall,
```

Comment:
They all write ```movabs r11``` instructions to different addresses. Generally, the asm code uses this value for something, before it is overwritten again.